### PR TITLE
Search pub

### PR DESCRIPTION
**Pull request by defong**
### Context
https://trello.com/c/Py1dedyJ/265-publish-a-single-course-to-search-via-a-json-endpoint

### Changes proposed in this pull request
The ability to publish courses via an authenticated api key endpoint

### CODE REVIEW

File: tests/Integration.Tests/Controllers/CoursesController.Tests.cs [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+            AssertOkay(result);
+
+            var resultingCourses = context.GetCoursesWithProviderSubjectsRouteAndCampuses()
+                .Include(x => x.DescriptionSections)
+                .ToList();
+
+            var resultingProviders = context.Providers.ToList();
+
+            // deduplicated
+            Assert.AreEqual(1, resultingProviders.Count);
```
Comment (timabell) [inline tests/Integration.Tests/Controllers/CoursesController.Tests.cs]: fluent assertions please :-)

File: src/api/Controllers/CoursesController.cs [hunk 2]
```diff
@@ -26,6 +28,49 @@ public CoursesController(ICourseDbContext courseDbContext)
...
+                    AssociateWithLocations(ref courses);
+
+                    _context.Courses.AddRange(courses);
+                    _context.SaveChanges();
+
+                    return Ok();
+                }
+                catch(Exception)
+                {
+                    return BadRequest();
```
Comment (timabell) [inline src/api/Controllers/CoursesController.cs]: log?

File: tests/Integration.Tests/Controllers/CoursesController.Tests.cs [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+using Microsoft.EntityFrameworkCore;
+using FluentAssertions;
+using Microsoft.Extensions.Logging;
+
+namespace GovUk.Education.SearchAndCompare.Api.Tests.Integration.Tests.Controllers
+{
+    [TestFixture]
+    public class CoursesControllerTests : CourseDbContextIntegrationBase
+    {
+        CoursesController subject;
```
Comment (timabell) [inline tests/Integration.Tests/Controllers/CoursesController.Tests.cs]: `subject` makes the tests harder to read.

File: tests/Unit.Tests/Client/HttpClientWrapperTests.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    [TestFixture]
+    public class HttpClientWrapperTests
+    {
+        private HttpClientWrapper sut;
+        private  Mock<HttpMessageHandler> mockMessageHandler;
+        [SetUp]
+        public void SetUp()
+        {
+            mockMessageHandler = new Mock<HttpMessageHandler>();
+            sut = new HttpClientWrapper(new HttpClient(mockMessageHandler.Object));
```
Comment (timabell) [inline tests/Unit.Tests/Client/HttpClientWrapperTests.cs]: sut sucks as a name imho, call it httpClientWrapper

File: tests/Integration.Tests/Controllers/CoursesController.Tests.cs (hunk 3 above)
Comment (defong) [reply tests/Integration.Tests/Controllers/CoursesController.Tests.cs]: disagree if you have a list of name items then what is the subject under test?

File: tests/Unit.Tests/Client/HttpClientWrapperTests.cs (hunk 4 above)
Comment (defong) [reply tests/Unit.Tests/Client/HttpClientWrapperTests.cs]: disgree that is the subject under test

**Pull request merged**